### PR TITLE
Keeping pipelines full

### PR DESCRIPTION
**Pull request by goliaro**
## Why are these changes needed?

These changes are needed to avoid over-requesting new workers when using pipelining to submit tasks from owners to their workers. When a new task is submitted to an owner, the code first tries to send the task to an existing worker if the number of tasks in flight to that worker is less than the maximum allowed by the pipelining settings. If all pipelines to all workers are full, then it requests a new worker.

## Checks

- [ x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ x] I've included any doc changes needed for https://docs.ray.io/en/latest/.
- [ x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: src/ray/core_worker/transport/direct_task_transport.h
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.h:223]: Discussed offline about refactoring this class to squash all of the fields that are currently related to `SchedulingKey` into one hashmap. The value should be a new struct that includes:
- these new counts
- the workers that are currently leased to us with that scheduling key (`worker_to_lease_entry_`)
- whether there is a pending lease request for that key (`pending_lease_requests_`)
- the tasks queued with that key (`task_queues_`)

**Comment by stephanie-wang** [comment]
I'll take another look once the refactoring and unit tests are done!

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:253]: I'm not sure about putting this logic in `RequestNewWorkerIfNeeded`. It seems a bit strange to be submitting tasks in this method (before the logic only had to do with requesting new workers and not anything to do with task submission). It also seems a bit brittle because `OnWorkerIdle`, called below, calls back into `RequestNewWorkerIfNeeded`.

Instead, how about we only use the check to see whether we should request a new worker or not? Then, we should also move this new logic to submit tasks to already active workers to when the task is first queued in `SubmitTask`.

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:267]: I'm not sure if we want to break here. Couldn't there be multiple idle workers that could get filled by new tasks? But I guess this depends on where we decide to put this logic (see above comment).

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:290]: I don't think this check is necessary (std::move should guarantee this already).

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:330]: I don't think this check is necessary. The previous check was just to make sure that there wasn't already a pending lease request for the same scheduling key (arguably also not necessary).

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:472]: I don't think this is necessary because we should still get back the callback for `PushNormalTask`.

File: src/ray/core_worker/transport/direct_task_transport.h
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.h:218]: Consider making this a hashmap from rpc::WorkerAddress -> LeaseEntry instead of keeping a separate hashmap.

**Review by stephanie-wang — CHANGES_REQUESTED** [review]

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.cc:253]: The reason why I added this part (calling OnWorkerIdle from within RequestNewWorkerIfNeeded) was to avoid introducing new latency due to the fact that `OnWorkerIdle` works in a _pull_ (rather than _push_) fashion. `OnWorkerIdle` is normally only called when (1) we get a new worker, or (2) when we get a response from a worker (after the worker has completed the execution of a task). At that point, we pull tasks from the queue and submit them.  
Now, consider a scenario with a set of active workers with non-full pipelines, where we have just added a few more tasks to a queue at the owner's. In this situation, even if some of the pipelines are not full, before we can submit the new tasks to an active worker, we have to wait until that worker has responded back to the owner (because OnWorkerIdle is not called until then). If we submit to a queue at the owner's a number of tasks that is larger than the grand total of existing "spots" available in the pipelines to the existing worker, the system will only fill those pipelines (and realize that it needs to request an additional worker) only after it has received a response from every single one of the existing workers with non-full pipelines. As a result, the total number of active workers will be lower for a longer period of time, and the overall execution time will suffer.

Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_task_transport.cc:253]: Yes, I understand that we should call `OnWorkerIdle` even while it is still executing other tasks. My comment was more about where we should call it. I think it is better to call it directly in `SubmitTask` right after we've added tasks to the queue. The reason is that: a) dispatching tasks doesn't match the current semantics of the method, which is supposed to only request a new worker, and b) it prevents the bad recursive structure where `RequestNewWorkerIfNeeded` calls `OnWorkerIdle` calls `RequestNewWorkerIfNeeded`, etc.

Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.cc:253]: Oh I see! I think I misread your initial comment! Sorry about that. This makes sense. So I guess we would just move the code block 

```cpp
if (scheduling_key_entry.tot_tasks_in_flight <
      scheduling_key_entry.active_workers_.size() * max_tasks_in_flight_per_worker_) {
    // The pipelines to the current workers are not full yet, so we don't need more
    // workers.

    // Find a worker with a number of tasks in flight that is less than the maximum
    // value (max_tasks_in_flight_per_worker_) and call OnWorkerIdle to send tasks to
    // that worker
    for (auto active_worker_addr : scheduling_key_entry.active_workers_) {
      RAY_CHECK(worker_to_lease_entry_.find(active_worker_addr) !=
                worker_to_lease_entry_.end());
      auto &lease_entry = worker_to_lease_entry_[active_worker_addr];
      if (lease_entry.tasks_in_flight_ < max_tasks_in_flight_per_worker_) {
        OnWorkerIdle(active_worker_addr, scheduling_key, false,
                     lease_entry.assigned_resources_);
        break;
      }
    }

    return;
  } 
```

(without the return statement, of course) to the `SubmitTask` function. `RequestNewWorkerIfNeeded` would then only keep the following if statement?

```cpp
if (scheduling_key_entry.tot_tasks_in_flight <
      scheduling_key_entry.active_workers_.size() * max_tasks_in_flight_per_worker_) {
    // The pipelines to the current workers are not full yet, so we don't need more
    // workers.

    return;
  } 
```

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.cc:267]: I see! So I guess we would remove the break statement if we put this in `SubmitTask`, right?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_task_transport.cc:253]: Yes, exactly!

Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.cc:253]: Sounds good! Let me do that right now.

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_task_transport.cc:267]: Hmm I need to think about that. It seems like we could structure the code so that we're always guaranteed that only one worker needs to be filled up during `SubmitTask`, but I'm not sure.

File: src/ray/core_worker/transport/direct_task_transport.h
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.h:218]: Doesn't `worker_to_lease_entry_` already have this mapping? Do you mean that I should just move `worker_to_lease_entry_` into the `SchedulingKeyEntry` struct, so that each SchedulingKey will be paired to its own `worker_to_lease_entry_` hashmap?

**Comment by stephanie-wang** [comment]
Hmm actually I think I'm just wrong about this one haha. Ignore that!

**Comment by stephanie-wang** [comment]
Yes! I'm not sure which would be cleaner. I'll leave that up to you.

**Comment by gabrieleoliaro** [comment]
@stephanie-wang  Ok, I just pushed the updated code :)

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (ffbin) [inline src/ray/core_worker/transport/direct_task_transport.cc:267]:
```suggestion
  // Check whether we really need a new worker or whether we have
```

File: src/ray/core_worker/transport/direct_task_transport.h
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.h:211]: Could you remove the tail underscore at the end of these field names? Usually we do this only for private class members (see [Google style guide](https://google.github.io/styleguide/cppguide.html#Variable_Names)). Realizing now that the above `LeaseEntry` struct doesn't follow this convention, oops :)

**Review by stephanie-wang — COMMENTED** [review]
A few more comments :)

File: src/ray/core_worker/transport/direct_task_transport.h
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.h:211]: Should I also remove the underscores from the field names in the LeaseEntry struct? I guess it's better late than never, right?

Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_task_transport.h:211]: That would be great, thanks! :)

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_task_transport.cc:80]: Hmm okay now that I'm thinking about this again, I think it is okay to have the `break` statement and only dispatch to one worker! Sorry about that :)

Could you add a comment explaining why it's okay to break after one worker?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.cc:186]: I just added the new method! However, I am not sure about attempting to delete at this place, because there would still be some worker in the active_workers set, so calling the new method would tell us that it's not safe to delete the entry yet

Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_task_transport.cc:186]: Ahh gotcha, makes sense!

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.cc:80]: Sounds good, let me add that right now!

**Comment by gabrieleoliaro** [comment]
Ok, I just pushed the updated code!

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:82]: Suggest wrapping this in a `const` method of the `SchedulingKeyEntry` (e.g., `HasAvailableWorkers()`) to make it more readable!

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:93]: Suggest wrapping this in a const method of the WorkerLeaseEntry to make it more readable!

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:266]: Shouldn't we check if it's okay to delete the `SchedulingKeyEntry` here too?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:378]: Do we need to check if we should delete the SchedulingKeyEntry here?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.cc:423]: Do we need to check whether we should delete the SchedulingKeyEntry here?

File: src/ray/core_worker/transport/direct_task_transport.h
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.h:223]:
```suggestion
    bool CanDelete() {
```

**Review by stephanie-wang — COMMENTED** [review]
Thanks, this is looking very close! I just had some questions about making sure we delete the SchedulingKeyEntry properly. It would also be good to add unit tests for that, to make sure we're not leaking any memory.

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.cc:423]: Yes! However, I think that the check should be placed inside the callback function in `CancelWorkerLeaseIfNeeded`, because the callback can call `CancelWorkerLeaseIfNeeded` as well, and `CancelWorkerLeaseIfNeeded` needs to access the SchedulingKeyEntry

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.cc]: I was wondering, what do you mean by a `const` method? Also, I needed to define the method outside of the `SchedulingKeyEntry` struct to be able to access the `max_tasks_in_flight_per_worker_` member of `CoreWorkerDirectTaskSubmitter`. Otherwise, the compiler reported an error. Do you know if there is a way to access `max_tasks_in_flight_per_worker_` from a method within the `SchedulingKeyEntry` struct?

**Comment by gabrieleoliaro** [comment]
> Thanks, this is looking very close! I just had some questions about making sure we delete the `SchedulingKeyEntry` properly. It would also be good to add unit tests for that, to make sure we're not leaking any memory.

I was wondering what type of unit test you had in mind to check that we are not leaking memory by forgetting to delete some entries in the `scheduling_key_entries` hashmap. Because the hashmap is a private field of the `CoreWorkerDirectTaskSubmitter` class, we can't just check its size from within `direct_task_transport_test.cc` . Should I add a public function that allows us to check the size?

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (stephanie-wang) [reply src/ray/core_worker/transport/direct_task_transport.cc]: Ah I just mean to mark that the method does not modify the `SchedulingKeyEntry`, like this:
`CanDelete() const { ... }`

Yes, it shouldn't be able to access it because it's a private member of `CoreWorkerDirectTaskSubmitter`. You can fix it by allowing `CanDelete` to take the max tasks as an argument.

**Comment by stephanie-wang** [comment]
Yes, you can either add a public function to check the size or you can make the unit test class a `friend` of the CoreWorkerDirectTaskSubmitter class (see [here](https://en.cppreference.com/w/cpp/language/friend)).

File: src/ray/core_worker/transport/direct_task_transport.cc
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.cc]: Sounds good! I'll add the const keyword, and allow CanDelete() (and the other two methods I added to check whether the pipeline fullness) to take `max_tasks_in_flight_per_worker_` as an argument, so that I can place the function inside the struct

File: src/ray/core_worker/transport/direct_task_transport.h
Comment (stephanie-wang) [inline src/ray/core_worker/transport/direct_task_transport.h:95]: You can also use `scheduling_key_entries_.empty()` here!

**Review by stephanie-wang — APPROVED** [review]
Looks great! I'll merge once Travis finishes.

File: src/ray/core_worker/transport/direct_task_transport.h
Comment (gabrieleoliaro) [reply src/ray/core_worker/transport/direct_task_transport.h]: Good call! Just changed this. I also removed the RAY_LOG(INFO) line

**Comment by stephanie-wang** [comment]
Hey @gabrieleoliaro, looks like there is a build error. Could you fix it? https://api.travis-ci.com/v3/job/378975005/log.txt

**Comment by stephanie-wang** [comment]
@gabrieleoliaro, I'm not sure if that last commit will fix the issue. That annotation means that the compiler should check that the mutex is held whenever that method is called. But the method is public and the mutex is private, so I don't think it will work.

You can fix the error by acquiring the lock inside the method!

**Comment by gabrieleoliaro** [comment]
> Hey @gabrieleoliaro, looks like there is a build error. Could you fix it? https://api.travis-ci.com/v3/job/378975005/log.txt

Just pushed!

**Comment by gabrieleoliaro** [comment]
> @gabrieleoliaro, I'm not sure if that last commit will fix the issue. That annotation means that the compiler should check that the mutex is held whenever that method is called. But the method is public and the mutex is private, so I don't think it will work.

sorry, this was just a first commit :)

**Comment by ffbin** [comment]
https://travis-ci.com/github/ray-project/ray/jobs/379015992 Hi @gabrieleoliaro , the java ci job failure is related to the pr, pls help take a look, thanks.

**Comment by stephanie-wang** [comment]
> https://travis-ci.com/github/ray-project/ray/jobs/379015992 Hi @gabrieleoliaro , the java ci job failure is related to the pr, pls help take a look, thanks.

Really? Seems unlikely, the test is failing on master too: https://travis-ci.com/github/ray-project/ray/jobs/379177286

**Comment by ffbin** [comment]
> > https://travis-ci.com/github/ray-project/ray/jobs/379015992 Hi @gabrieleoliaro , the java ci job failure is related to the pr, pls help take a look, thanks.
> 
> Really? Seems unlikely, the test is failing on master too: https://travis-ci.com/github/ray-project/ray/jobs/379177286

Sorry, It should be irrelevant.

**Pull request merged**
